tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_packing.cc

```diff
@@ -0,0 +1,476 @@
+    size_t maxUserangeId =
+        computeAllocationInfos(allocInfos, userangeAnalysis, allocs);
+
+    // Sort the allocation infos.
+    std::sort(allocInfos.begin(), allocInfos.end(), compare);
+
+    size_t optMemory = 0;
+    size_t currentOffset = 0;
+    size_t globalBufferOffset = 0;
+
+    for (auto currentIter = allocInfos.begin(); currentIter != allocInfos.end();
+         ++currentIter) {
+      size_t offsetInBytes = currentOffset * 64;
+      UseInterval::Vector useranges = *(currentIter->userangeIntervals);
+
+      // Compute userange gaps.
+      std::vector<std::pair<UseInterval, size_t>> gaps =
+          currentIter->computeGaps(maxUserangeId);
+
+      // Add the current buffer to the packing infos.
+      packingInfos.emplace_back(currentIter->alloc, bufferId, offsetInBytes);
+      if (gaps.empty())
+        continue;
+
+      for (auto checkedAllocInfoIter = std::next(currentIter);
+           checkedAllocInfoIter != allocInfos.end();) {
+
+        // Check if a gap exists to pack the memory into.
+        // If not continue.
+        if (!findGapAndUpdate(gaps, packingInfos, *checkedAllocInfoIter,
+                              *currentIter, offsetInBytes)) {
+          ++checkedAllocInfoIter;
+          continue;
+        }
+        optMemory += checkedAllocInfoIter->numSegments;
+        checkedAllocInfoIter = allocInfos.erase(checkedAllocInfoIter);
+      }
+      // Increase the total global memory offset.
+      currentOffset += currentIter->numSegments;
+    }
+
+    size_t totalBufferSize = currentOffset * 64;
+    bufferGen.insert(std::make_pair(bufferId, totalBufferSize));
+  }
+
+private:
+  const size_t windowSize;
+  const size_t bufferId = 0;
+  const CompareT compare;
+
+  /// We try to find an appropriate userange gap to pack the buffer into it.
+  /// If we find one we update only the gaps and the buffer offset map.
+  bool findGapAndUpdate(std::vector<std::pair<UseInterval, size_t>> &gaps,
+                        std::vector<PackedInfo> &allocBufferOffsets,
+                        AllocationInfo &allocToPack,
+                        AllocationInfo &allocToPackInto,
+                        size_t unitedBufferOffset) {
+    // Check if the buffer to pack into has enough memory.
+    if (allocToPackInto.numSegments < allocToPack.numSegments)
+      return false;
+    for (auto gapIter = gaps.begin(); gapIter != gaps.end();) {
+      size_t gapUserangeSize = computeUserangeSize(gapIter->first);
+
+      // If a gap interval has no free contiguous memory anymore, erease it from
+      // list.
+      if (gapIter->second <= 0) {
+        gapIter = gaps.erase(gapIter);
+        continue;
+      }
+
+      // Checks if enough the userange and contigous memory segments is
+      // free.
+      if (gapIter->second < allocToPack.numSegments ||
+          allocToPack.firstUse < gapIter->first.start ||
+          allocToPack.lastUse > gapIter->first.end) {
```

Comment:
If you keep the gaps sorted (which they initially are) you could abort here once the gap is past the end of the alloc.